api/api.go

```diff
@@ -93,6 +93,9 @@ type WriteOptions struct {
 	// ctx is an optional context pass through to the underlying HTTP
 	// request layer. Use Context() and WithContext() to manage this.
 	ctx context.Context
+
+	// IdempotencyToken can be used to ensure the write is idempotent.
+	IdempotencyToken string
```

Comment:
I'm conflicted about having the field in WriteOptions and slightly lean towards how you had it before under `JobDispatchRequest`. Idempotency is such a common concern, so belonging to `WriteOptions` make sense. Though, it currently only applies to dispatch jobs; other APIs address a similar but not an exact problem using CAS and ModifyIndex (e.g. [Update Scheduler Configuration endpoint](https://www.nomadproject.io/api-docs/operator/scheduler#update-scheduler-configuration)).

I'm fine leaving it here, as this is probably the path forward - but would recommend calling out that only the dispatch endpoint supports it.